keps/sig-node/4569-cgroup-v1-deprecation/README.md

```diff
@@ -0,0 +1,677 @@
+
+Users depending following technologies will have to use their respective listed versions below:
+
+- OpenJDK / HotSpot: jdk8u372, 11.0.16, 15 and later
+
+- NodeJs 20.3.0 or later
+
+- IBM Semeru Runtimes: jdk8u345-b01, 11.0.16.0, 17.0.4.0, 18.0.2.0 and later
+
+- IBM SDK Java Technology Edition Version (IBM Java): 8.0.7.15 and later
+
+- If users run any third-party monitoring and security agents that depend on the cgroup file system, they need to update the agents to a version that supports cgroup v2.
+
+## Design Details
+
+## Handling Deprecation Warnings and Migration Path Flexibility
+
+### Implementing Warning Logs for cgroup v1 Usage
+
+#### Background
+
+The Kubernetes kubelet relies on the [IsCgroup2UnifiedMode()](https://github.com/opencontainers/runc/blob/5e0ec3fbbff07ef4016c04ae584f18b27bbcfdc1/libcontainer/cgroups/utils.go#L34C6-L34C26) function from the `libcontainer` package within `runc` to ascertain the cgroup version the host system is using.
+
+#### Introduction of DeprecatedCgroupV1 Feature Gate
+
+With the introduction of the `DeprecatedCgroupV1` feature gate, Kubernetes aims to guide users through the transition from cgroup v1 to cgroup v2 by providing clear and timely warnings. When this feature gate is enabled, kubelet will log a warning message in scenarios where it detects the host system is utilizing cgroup v1. This approach ensures users are informed of the deprecation and encourages planning for migration to cgroup v2.
+
+
+#### Code Modifications for Warning Log Implementation
+
+To incorporate these warning logs, the kubelet's code will be adjusted in locations where the IsCgroup2UnifiedMode() function is invoked. e.g.
+
+Original Code Snippet:
+```golang
+if m.memorySwapBehavior == kubelettypes.LimitedSwap {
+		if !isCgroup2UnifiedMode() {
+			swapConfigurationHelper.ConfigureNoSwap(lcr)
+			return
+		}
+	}
+```
+Revised Code Snippet:
+```golang
+if m.memorySwapBehavior == kubelettypes.LimitedSwap {
+    if !isCgroup2UnifiedMode() && utilfeature.DefaultFeatureGate.Enabled(features.DeprecatedCgroupV1) {
+      klog.Warning("cgroup v1 support has been deprecated, please plan for the migration towards cgroup v2")
+      swapConfigurationHelper.ConfigureNoSwap(lcr)
+      return
+    }
+}
+```
+
+Furthermore, code segments that default to cgroup v1 logic will be inverted to assume cgroup v2 as the default. This shift underscores the transition to cgroup v2 and ensures warning logs are consistently generated for cgroup v1 usage. e.g.
+
+Original Code Snippet:
+```golang
+	memLimitFile := "memory.limit_in_bytes"
+	if libcontainercgroups.IsCgroup2UnifiedMode() {
+		memLimitFile = "memory.max"
+	}
+```
+Revised Code Snippet:
+```golang
+	memLimitFile = "memory.max"
+	if libcontainercgroups.!IsCgroup2UnifiedMode() && utilfeature.DefaultFeatureGate.Enabled(features.DeprecatedCgroupV1) {
```

Comment:
If we go with a kubelet flag than I think we would need both kubelet flag and feature gate?

Reply:
Yeah, you are right, missed that part. Thanks, I will update it.